src/huggingface_hub/_commit_api.py

```diff
@@ -395,13 +395,15 @@ def prepare_commit_payload(
     deletions: Iterable[CommitOperationDelete],
     commit_message: str,
     commit_description: Optional[str] = None,
+    parent_commit: Optional[str] = None,
 ):
     """
     Builds the payload to POST to the `/commit` API of the Hub
     """
     commit_description = commit_description if commit_description is not None else ""
 
-    return {
+    payload = {
+        **({"parentCommit": parent_commit} if parent_commit is not None else {}),
```

Comment:
I saw that this line has been added via this commit https://github.com/huggingface/huggingface_hub/pull/916/commits/9ff8951a5939b04f576911edd4d20ac06fa5140c with commit message `parentCommit needs to be at top of the payload`.

Since it's a quite unusual syntax (`**({"parentCommit": parent_commit} if parent_commit is not None else {}),`) I would definitely put a comment here to prevent someone from changing that without knowing. Also, it's using the fact that Python dictionaries are ordered by default which seems more a "desirable side-effect" here than an explicit intention from the developer.

---
To prevent any future mistakes, I would even be tempted to have an helper function `_reorder_payload` that explicitly does the job just before sending to the API. Something like:

```py
def _reorder_payload(payload: Dict[str, Any], data_fields: List[str]) -> Dict[str, Any]:
    """Reorder keys in payload in order to have metadata information at the beginning.

    Note: I am not sure what's best between an explicit `data_fields` or an explicit
          `metadata_fields`. I tend to think that metadata fields will evolve
          more often so having them ordered implicitly is better.
    """
    ordered_payload: Dict[str, Any] = dict()
    
    # First, add metadata fields
    for key, value in payload.items():
        if key not in data_fields:
            ordered_payload[key] = value
    
    # Then, add data fields
    for key in data_fields:
        if key in payload:
            ordered_payload[key] = payload[key]
            
    return ordered_payload
``` 

And then use it just before the `return payload`. Something like:

```py
def prepare_commit_payload(...):
    (...)
    # build the payload without any concerns
    (...)
    return _reorder_payload(payload, data_fields=["files", "lfsFiles", "deletedFiles"])
```

Reply:
actually - see https://github.com/huggingface/moon-landing/pull/3528#issuecomment-1206597790 - we're going to use a different content-type for JSON streaming, so the restriction on payload order is moot

Reply:
Ok thanks for pinging me on this one. Then for here just a comment is fine I guess.